ic: implement proxy mode

pivotal-ning-yu opened the pull request:
By setting gp_interconnect_proxy to 'proxy' now the interconnect could
run in the proxy mode.  In the proxy mode a proxy bgworker is launched
on each segment, the proxy bgworkers connect to each other, the backends
only connect to the bgworker on their own segment.  By doing this we
reduced the total network flows, we also reduced the required port
counts.

To enable the proxy mode we need to first configure the guc
gp_interconnect_proxy_addresses, for example:

    gpconfig \
      -c gp_interconnect_proxy_addresses \
      -v "'1:-1:10.0.0.1:2000,2:0:10.0.0.2:2001,3:1:10.0.0.3:2002'" \
      --skipvalidation

----

The ic-proxy mode is disabled in configure by default, enable it manually with `--enable-ic-proxy`.

----

Refer to the ic-proxy [README](https://github.com/greenplum-db/gpdb/blob/b1eb3026ee336260ea0380eafa26fbcb216b1d40/src/backend/cdb/motion/README.ic-proxy.md) for details.

----

Below is a simple __bash__ script to set this automatically:

    #!/usr/bin/env bash

    portbase=2000

    psql -tqA -d postgres -P pager=off -F ' ' \
        -c "select dbid, content, $portbase+dbid as port, address from gp_segment_configuration order by 1" \
    | while read -r dbid content port addr; do
        ip=$(host "$addr" | fgrep -v IPv6 | head -n1)
        ip=${ip##* }
        echo "$dbid:$content:$ip:$port"
      done \
    | paste -sd, - \
    | xargs -rI'{}' gpconfig --skipvalidation -c gp_interconnect_proxy_addresses -v "'{}'"

It's only verified on ubuntu linux.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

danielgustafsson commented:
Please add the appropriate linker steps to autoconf such that the pipelines can actually build the code.

pivotal-ning-yu commented:
> Please add the appropriate linker steps to autoconf such that the pipelines can actually build the code.

Thanks for the suggestion, yes, we definitely will do this, it's in the TODO list.  It's not contained in the PR yet because so far we focused only on the code structure, there are still a lot of things to do before it's production-ready.

danielgustafsson commented:
> On 26 Mar 2020, at 05:45, Ning Yu <notifications@github.com> wrote:

> Thanks for the suggestion, yes, we definitely will do this, it's in the TODO list. It's not contained in the PR yet because so far we focused only on the code structure, there are still a lot of things to do before it's production-ready.

I get that, but it helps the rest of us looking on from the sidelines if the CI can at least try to build the code.

pivotal-ning-yu commented:
> On 26 Mar 2020, at 05:45, Ning Yu ***@***.***> wrote: Thanks for the suggestion, yes, we definitely will do this, it's in the TODO list. It's not contained in the PR yet because so far we focused only on the code structure, there are still a lot of things to do before it's production-ready.
> I get that, but it helps the rest of us looking on from the sidelines if the CI can at least try to build the code.

Good point.  I'll do that today.

pivotal-ning-yu commented:
Force-pushed, corrected the commit message.

Also pushed commits to fix the compilation, and detect libuv with `configure`.

danielgustafsson commented:
> Also pushed commits to fix the compilation, and detect libuv with `configure`.

Thanks!

To get this building in Travis, you would something like the below (untested) diff:

```patch
diff --git a/.travis.yml b/.travis.yml
index c339923946..9b13253876 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -20,12 +20,16 @@ addons:
             - libxml2-dev
             - libevent-dev
             - libperl-dev
+            - libuv1-dev
             - g++-8
             - python-dev
             - python-yaml
             - libapr1-dev
             - libzstd1
             - libzstd1-dev
+    homebrew:
+        packages:
+            - libuv

 matrix:
     include:
```

pivotal-ning-yu commented:
> To get this building in Travis, you would something like the below (untested) diff:

Cool, thanks, pushed to this PR, let's wait and see.

pivotal-ning-yu commented:
Force-pushed.  To let interconnect use the proxy, we used to base our logic on-top of ic-tcp, we reused the ic-tcp socket operations to communicate through the domain sockets.  However to replace the domain sockets with shared memory, we need to get rid of these plain socket operations, so we moved the logic to a ic-proxy specific source file, moved the socket operations to an abstract layer.  The code is buggy on EOS handling, as a result some packets can be lost during the query.  This is not a big matter, as we will now begin the switching to shared memory, it will have different EOS handling logic.

pivotal-ning-yu commented:
Pushed 2 commits to optimize the previous domain socket based approach.  The key point is to set the `TCP_NODELAY` option, so the kernel won't merge the sending requests for us, which causes very bad performance in OLTP queries.  On the other hand, we merge small requests by ourselves, so it makes no sense to let the kernel do it again.

Also optimized the memory usage, and fixed some memory leaks, at the moment the ic-proxy OLTP performance is about 80% of ic-tcp in my local env.  I also plan to verify it with a large cluster.

soumyadeep2007 commented:
w/ @d :
We couldn't find the setting of the `TCP_NODELAY`. Did you have a local change that you didn't git add?

pivotal-ning-yu commented on src/backend/cdb/motion/ic_proxy_server.c:
```diff
@@ -0,0 +1,375 @@
+	ic_proxy_log(LOG, "setup peer listener on port %d", IC_PROXY_PORT);
+
+	uv_tcp_init(loop, peer_listener);
+	uv_tcp_nodelay(peer_listener, true);
```
> w/ @d :
> We couldn't find the setting of the `TCP_NODELAY`. Did you have a local change that you didn't git add?

@soumyadeep2007 The setting is made via the libuv API, `uv_tcp_nodelay()`.

pivotal-ning-yu commented on src/backend/cdb/motion/ic_proxy_server_peer.c:
```diff
@@ -0,0 +1,476 @@
+	ic_proxy_pktbuf_init(&peer->pktbuf);
+
+	uv_tcp_init(loop, &peer->tcp);
+	uv_tcp_nodelay(&peer->tcp, true);
```
@soumyadeep2007 as well as here, it's set on the proxy-proxy tcp sockets.

pivotal-ning-yu commented:
Since the commit `server: let the primary & mirror proxies listen on different ports` we need a different syntax for the GUC `gp_interconnect_proxy_addresses`, please refer to the PR description for details.

pivotal-ning-yu commented:
So far, we have passed most of the ICW tests on a devel pipeline, there are still problems with mirror promotion and standby activation.

pivotal-ning-yu commented:
Force pushed, we rebased and squashed the commits, it's time to remove the WIP flag, the PR is ready for review.  The changes for the devel pipeline were not removed from the PR, we need to adjust the changes to better fit the master pipelines.  During the review we will still make some minor improvements, such as error handling, documents, and comments.

pivotal-ning-yu commented:
Pushed a commit to re-enable the ICW pipeline jobs.

zhangh43 commented:
I'm reviewing this PR.

zhangh43 commented on src/backend/cdb/motion/ic_proxy_client.c:
```diff
@@ -0,0 +1,1170 @@
+			ICProxyPkt *pkt = lfirst(cell);
+
+			ic_proxy_log(WARNING, "%s: unhandled cached %s, dropping it",
+						 ic_proxy_client_get_name(client),
```
is it just a warning? will it lose some packets and leads to wrong answer?

pivotal-ning-yu replied:
Yes, it's just a warning.  It should never happen in practice, but once it happens we should know it.  Maybe `ERROR` is better than `WARNING` here.

pivotal-ning-yu commented on src/backend/cdb/motion/ic_proxy.h:
```diff
@@ -0,0 +1,58 @@
+#define IC_PROXY_LOG_LEVEL LOG
+#endif
+
+//#define ic_proxy_alloc(size) MemoryContextAlloc(CurrentMemoryContext, size)
```
+1, they were only used during the early development; retired in a new commit.

pivotal-ning-yu commented on src/backend/cdb/motion/ic_proxy_addr.c:
```diff
@@ -0,0 +1,123 @@
+		ic_proxy_addrs = NIL;
+	}
+
+	ic_proxy_content_count = -2;
```
Updated by following your suggestions.

pivotal-ning-yu commented on src/backend/cdb/motion/ic_proxy_addr.c:
```diff
@@ -0,0 +1,123 @@
+	{
+		ICProxyAddr *addr = lfirst(cell);
+
+		if (addr->dbid == dbid)
```
A mirror has different dbid with its primary, and the dbid is unchanged during mirror promotion.  I guess you mean segid?

On the other hand, mirrors and standby are included in the `gp_interconnect_proxy_addresses`, my example in the PR description is confusing, I should update it.

pivotal-ning-yu commented on src/backend/cdb/motion/ic_proxy_addr.h:
```diff
@@ -0,0 +1,41 @@
+ * List<ICProxyAddr *>
+ */
+extern List		   *ic_proxy_addrs;
+extern int			ic_proxy_content_count;
```
Nice catch, retired.  It was used outside during the development, but the user has retired already.

pivotal-ning-yu commented on src/backend/cdb/motion/ic_proxy_backend.c:
```diff
@@ -0,0 +1,119 @@
+	ic_proxy_build_server_sock_path(addr.sun_path, sizeof(addr.sun_path));
+
+	fd = socket(AF_UNIX, SOCK_STREAM, 0);
+	Assert(fd != -1);
```
Updated.

pivotal-ning-yu commented on src/backend/cdb/motion/ic_proxy_backend.c:
```diff
@@ -0,0 +1,119 @@
+
+/*
+ * FIXME: this function is to make the debugging easier, however as it
+ * establishes the connection, as well as the registeration, one by one, it has
```
It's not a trivial change.  When we decide to do this we should first move the proxy logic outside ic-tcp.c, we need to rewrite all the backend logic for ic-proxy, such as:
- send data with the proxy pkt header instead of the tcp header;
- connect / shutdown / recv / send with libuv;

We'd better do it as a separate story.  BTW, per my tpc-h benchmark, the current one-by-one hand shaking does not introduce a significant latency, because we only hand shake with the local proxy.  If we need to do a global hand shaking, that will cause a huge latency, and we should consider the parallel style and above refactor.

pivotal-ning-yu commented on src/backend/cdb/motion/ic_proxy_bgworker.c:
```diff
@@ -0,0 +1,59 @@
+ICProxyStartRule(Datum main_arg)
+{
+#if 0
+	/* we only start fts probe on master when -E is specified */
```
Ah, yes, the code is unused.  Removed in a new commit.

pivotal-ning-yu commented on src/backend/cdb/motion/ic_proxy_bgworker.c:
```diff
@@ -0,0 +1,59 @@
+	 * reread postgresql.conf if requested
+	 */
+	pqsignal(SIGHUP, sigHupHandler);
+	pqsignal(SIGINT, sigIntHandler);
```
In fact we could receive the cancel signal already.  Anyway I have pushed a commit to add an explicit `SIGINT` handler in the libuv way.

pivotal-ning-yu commented on src/backend/cdb/motion/ic_proxy_client.c:
```diff
@@ -0,0 +1,1170 @@
+							 ic_proxy_client_get_name(client));
+
+			placeholder->successor = client;
+			return;
```
> it's more clear to return REGISTER_DELAYED, and handle in caller `ic_proxy_client_on_sent_hello_ack`

The caller knows the register is delayed by checking the `state` bits.

> BTW, if delay happens, when will `ic_proxy_client_on_sent_hello_ack` be triggered again ?

Note that we mark the `client` as a `successor`, so once the `placeholder` is unregistered in `ic_proxy_client_unregister()`, it will register its `successor`.

kmacoskey commented:
A couple of observations:

The changes install packages on every execution of the implemented tests. The result is that every commit, in every production pipeline, and every development pipeline, will attempt to make a connection to an upstream URL to download bits. This impacts the flakiness of those tests, execution time, and the strain on the CI infrastructure. There are established methods to avoid these problems, @edespino and @greenplum-db/gp-releng can help with this.

Instead of the configure flag being `--without-libuv`/`--with-libuv`, would it be better to do `--enable-ic-proxy`/`--disable-ic-proxy` so that it's clear that the feature itself is what has a dependency on libuv?

Should `--with-libuv` be the default in the configuration when the necessary dependency is not officially supported on most of the platforms Greenplum supports?

pivotal-ning-yu commented:
> A couple of observations:
> 
> The changes install packages on every execution of the implemented tests. The result is that every commit, in every production pipeline, and every development pipeline, will attempt to make a connection to an upstream URL to download bits. This impacts the flakiness of those tests, execution time, and the strain on the CI infrastructure. There are established methods to avoid these problems, @edespino and @greenplum-db/gp-releng can help with this.

Yes, that will be cool to prevent those pre-setup.

> 
> Instead of the configure flag being `--without-libuv`/`--with-libuv`, would it be better to do `--enable-ic-proxy`/`--disable-ic-proxy` so that it's clear that the feature itself is what has a dependency on libuv?

+1, I'll make this change to this PR.

> 
> Should `--without-libuv` be the default in the configuration because the necessary dependency is not officially supported on most of the platforms Greenplum supports?

Makes sense, I'll change it.

pivotal-ning-yu commented:
Pushed a commit to rename the option to `--enable-ic-proxy`, the default setting is `false`, so since now on to compile with ic-proxy we must:
1. install `libuv`;
2. `configure --enable-ic-proxy`;

zhangh43 commented on src/backend/cdb/motion/README.ic-proxy.md:
```diff
@@ -0,0 +1,619 @@
+
+It is not really a problem to have two connections between two segments, but we
+could solve it easily: segment I only connects to segment J if `I < J`.  It is
+simply enough and is what we used initially, but is it good enough?
```
better to also illustrate how to handle failover and gpexpand case. We use a timer right?

zhangh43 commented on src/backend/cdb/motion/README.ic-proxy.md:
```diff
@@ -0,0 +1,619 @@
+   be able to establish the connections to all the other segments, too,
+   furthermore, the connections with the previous master proxy must be replaced
+   with the standby ones;
+
```
Could you please describe how failover and gpexpand works in current proxy design?

pivotal-ning-yu replied:
Added sections in README on this two topics.

zhangh43 commented on src/backend/cdb/motion/ic_proxy.h:
```diff
@@ -0,0 +1,58 @@
+ * Every proxy on the same host must use a different path, this is important to
+ * let proxies from different segments or even different clusters to coexist.
+ *
+ * This is ensured by including the postmaster pid in the path.
```
PostPortNumber is not postmaster pid

pivotal-ning-yu replied:
Nice catch, fixed.

zhangh43 commented on src/backend/cdb/motion/ic_proxy_addr.c:
```diff
@@ -0,0 +1,130 @@
+		/*
+		 * format: dbid:segid:ip:port
+		 */
+		while (fscanf(f, "%d:%d:%[0-9.]:%d,", &dbid, &content, ip, &port) == 4)
```
what's the behavior if format of `gp_interconnect_proxy_addresses` is not correct?
When I config this GUC with wrong format and restart cluster, the cluster cannot start successfully. When I try to correct it, gpconfig failed due to cluster is down. So we'd better abort the gpconfig when format is not correct.
I also leave comments in the GUC definition.

pivotal-ning-yu replied:
Pushed a commit to raise a warning instead of an error on invalid setting of this GUC.  I think it's reasonable to add a check in `gpconfig`, but let's do it after we deciding to use such a GUC forever.  Personally I'm still not satisfied with such a GUC based solution.

zhangh43 commented on src/backend/utils/misc/guc_gp.c:
```diff
@@ -4345,6 +4348,19 @@ struct config_string ConfigureNamesString_gp[] =
+		},
+		&gp_interconnect_proxy_addresses,
+		"",
+		NULL, NULL, NULL
```
better to have check/assign hook of this GUC to check format.

pivotal-ning-yu replied:
> what's the behavior if format of `gp_interconnect_proxy_addresses` is not correct?
> When I config this GUC with wrong format and restart cluster, the cluster cannot start successfully. When I try to correct it, gpconfig failed due to cluster is down. So we'd better abort the gpconfig when format is not correct.
> I also leave comments in the GUC definition.

This simple parser will parse as many as it can, if there are syntax errors some proxy workers could not know their listener ports and will run into repeatedly retry.  But the cluster is still live, you can run DDLs, you can run calalog queries, you can run data queries in other ic modes, only data queries in ic-proxy mode will block.

> better to have check/assign hook of this GUC to check format.

Hmm, the guc hooks might cause worse behavior for this guc.

Without the guc hooks we could still launch the cluster and run queries in other ic modes, and we could correct the setting with `gpconfig`; With the guc hooks the cluster can not be launched on the syntax errors, we have to modify the `postgresql.conf` files on all the segments to correct the setting.

zhangh43 commented on src/backend/cdb/motion/ic_proxy_backend.c:
```diff
@@ -0,0 +1,120 @@
+		if (ret == -1 && errno == EINTR)
+			continue;
+
+		if (ret == -1)
```
this is the domain socket, if the socket file `/tmp/.s.proxy.xxx` is not cleaned, e.g. `pkill -9 postgres` what will happens here? will it result in infinite loop?

pivotal-ning-yu replied:
Yes, it will run into infinite loop if the domain socket file is missing or no one is listening on the other side.  The long-term solution would still be creating a new libuv style backend instead of the current ic-tcp based one.  But I think we could at least prevent this infinite loop by checking the interrupts for now.  I would push a patch to do this.

zhangh43 commented on src/backend/cdb/motion/ic_proxy_bgworker.c:
```diff
@@ -0,0 +1,34 @@
+/*-------------------------------------------------------------------------
+ *
+ * ic_proxy_bgworker.c
+ *	  TODO file description
```
Please also add TODO file description.

pivotal-ning-yu replied:
Done.

zhangh43 commented on src/backend/cdb/motion/ic_proxy_iobuf.c:
```diff
@@ -0,0 +1,281 @@
+static uint16
+ic_proxy_ibuf_get_packet_size_b2c(const void *data)
+{
+	return *(const uint32 *) data;
```
why don't b2c return packet length as below?
```
const ICProxyPkt *pkt = data;
return pkt->len;
```

pivotal-ning-yu replied:
Yes, comments are added to explain this.

zhangh43 commented on src/backend/cdb/motion/ic_proxy_key.c:
```diff
@@ -0,0 +1,111 @@
+}
+
+void
+ic_proxy_key_from_pid(ICProxyKey *key,
```
`ic_proxy_key_from_pid ` the function name is confusing, maybe `ic_proxy_set_key`?

pivotal-ning-yu replied:
Good suggestion, renamed to `ic_proxy_key_init()`.

zhangh43 commented on src/backend/cdb/motion/ic_proxy_client.c:
```diff
@@ -0,0 +1,1405 @@
+					 ic_proxy_client_get_name(client),
+					 ic_proxy_pkt_to_str(pkt));
+		/* TODO: callback? */
+		ic_proxy_pkt_cache_free(pkt);
```
It seems that we only call `callback` for loopback packet. Should we call `callback` in this function?

pivotal-ning-yu replied:
`callback` should be called for all `message`s; this also applies for loopback packets.  In general a `callback` is only provided for `message`s, this is ensured by the callers.

However whether should we call it here?  It only reaches here if the packet is routed to the wrong client, this should never happen, in fact this logic is surrounded by a `#ifdef USE_ASSERT_CHECKING` block.  In fact we could also convert this logic to an `Assert()`.

zhangh43 commented on src/backend/cdb/motion/ic_proxy_router.c:
```diff
@@ -0,0 +1,304 @@
+		 * reentrantable, which happens on PAUSE & RESUME messages, so we must
+		 * schedule it in a libuv check callback.
+		 *
+		 * TODO: when callback is NULL, we could pass the packet immediately.
```
Not quite understand here. 
1. the loopback here is two QEs(QE1 and QE2) which are on the same segment, but different gangs. So a motion node connect QE1 and QE2, e.g. QE1 send data to QE2. Then QE1 send data to proxy and proxy route data back to QE2. is that correct? Can you explain more about `ic_proxy_client_on_p2c_data ` is not reentrantable for the case of PAUSE & RESUME message. I think these two messages are both sent from QE2(data recevier) to QE1(data sender), right? 
2. `libuv check callback` refer to `ic_proxy_router_loopback_on_check` instead of the function arg `ic_proxy_sent_cb callback` right?

pivotal-ning-yu replied:
I hacked the code locally with below changes:
- send immediately w/o a delayed loopback request;
- panic in i/o bufs when being nest called;

Below is the callback on the panic:

```
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007f468a9d8859 in __GI_abort () at abort.c:79
#2  0x0000562f61708a32 in errfinish (dummy=0) at elog.c:749
#3  0x0000562f6170baaf in elog_finish (elevel=22, fmt=0x562f61aa14c8 "ic-proxy-ibuf: nested call") at elog.c:1751
#4  0x0000562f617cc98e in ic_proxy_obuf_push (obuf=0x7f46876aa970, data=0x0, size=0, callback=0x562f617c6992 <ic_proxy_client_route_c2p_data>, opaque=0x7f46876aa840) at ic_proxy_iobuf.c:341
#5  0x0000562f617c8843 in ic_proxy_client_maybe_pause (client=0x7f46876aa840) at ic_proxy_client.c:1382
#6  0x0000562f617c7e64 in ic_proxy_client_on_p2c_message (client=0x7f46876aa840, pkt=0x562f656e3fa0, callback=0x562f617c7c78 <ic_proxy_client_on_sent_c2p_pause>, opaque=0x7f46876aae88) at ic_proxy_client.c:1101
#7  0x0000562f617c820a in ic_proxy_client_on_p2c_data (client=0x7f46876aa840, pkt=0x562f656e3fa0, callback=0x562f617c7c78 <ic_proxy_client_on_sent_c2p_pause>, opaque=0x7f46876aae88) at ic_proxy_client.c:1191
#8  0x0000562f617cac5f in ic_proxy_router_route (loop=0x562f61da27c0 <ic_proxy_server_loop>, pkt=0x562f656e3fa0, callback=0x562f617c7c78 <ic_proxy_client_on_sent_c2p_pause>, opaque=0x7f46876aae88) at ic_proxy_router.c:229
#9  0x0000562f617c86b5 in ic_proxy_client_maybe_request_pause (client=0x7f46876aae88) at ic_proxy_client.c:1331
#10 0x0000562f617c81f0 in ic_proxy_client_on_p2c_data (client=0x7f46876aae88, pkt=0x562f6577e5a0, callback=0x0, opaque=0x0) at ic_proxy_client.c:1187
#11 0x0000562f617cac5f in ic_proxy_router_route (loop=0x562f61da27c0 <ic_proxy_server_loop>, pkt=0x562f6577e5a0, callback=0x0, opaque=0x0) at ic_proxy_router.c:229
#12 0x0000562f617c6a5f in ic_proxy_client_route_c2p_data (opaque=0x7f46876aa840, data=0x562f65714ba0, size=8156) at ic_proxy_client.c:429
#13 0x0000562f617ccb35 in ic_proxy_obuf_push (obuf=0x7f46876aa970, data=0x562f657826a0 "\274\037", size=8124, callback=0x562f617c6992 <ic_proxy_client_route_c2p_data>, opaque=0x7f46876aa840) at ic_proxy_iobuf.c:369
#14 0x0000562f617c6af0 in ic_proxy_client_on_c2p_data_pkt (opaque=0x7f46876aa840, data=0x562f657826a0, size=8124) at ic_proxy_client.c:448
#15 0x0000562f617cc781 in ic_proxy_ibuf_push (ibuf=0x7f46876aa988, data=0x562f657826a0 "\274\037", size=8124, callback=0x562f617c6a62 <ic_proxy_client_on_c2p_data_pkt>, opaque=0x7f46876aa840) at ic_proxy_iobuf.c:193
#16 0x0000562f617c6dc5 in ic_proxy_client_on_c2p_data (stream=0x7f46876aa840, nread=8124, buf=0x7ffdc6b3b570) at ic_proxy_client.c:523
#17 0x00007f468aea8761 in ?? () from /usr/lib/x86_64-linux-gnu/libuv.so.1
#18 0x00007f468aea9298 in ?? () from /usr/lib/x86_64-linux-gnu/libuv.so.1
#19 0x00007f468aeaee70 in uv.io_poll () from /usr/lib/x86_64-linux-gnu/libuv.so.1
#20 0x00007f468ae9e6dc in uv_run () from /usr/lib/x86_64-linux-gnu/libuv.so.1
#21 0x0000562f617c5f36 in ic_proxy_server_main () at ic_proxy_main.c:430
#22 0x0000562f617c51f2 in ICProxyMain (main_arg=0) at ic_proxy_bgworker.c:33
#23 0x0000562f614a2276 in StartBackgroundWorker () at bgworker.c:747
#24 0x0000562f614b7d86 in do_start_bgworker (rw=0x562f61fb5fc0) at postmaster.c:6129
#25 0x0000562f614b8127 in maybe_start_bgworker () at postmaster.c:6365
#26 0x0000562f614b42ca in reaper (postgres_signal_arg=17) at postmaster.c:3300
#27 <signal handler called>
#28 0x00007f468aacaf4a in __GI___select (nfds=6, readfds=0x7ffdc6b3fb80, writefds=0x0, exceptfds=0x0, timeout=0x7ffdc6b3faf0) at ../sysdeps/unix/sysv/linux/select.c:41
#29 0x0000562f614b1a46 in ServerLoop () at postmaster.c:1905
#30 0x0000562f614b1031 in PostmasterMain (argc=5, argv=0x562f61f85a20) at postmaster.c:1531
#31 0x0000562f6138f1d5 in main (argc=5, argv=0x562f61f85a20) at main.c:240
```

Note that `ic_proxy_obuf_push()` is nest called, so the buf pointer is damaged, if we do not panic there will be unexpected results, such as packet/data corruption and runtime errors.

By delaying all the loopback packets the i/o bufs will not be nest called.  However an optimization can be made: first we still need to track the nest level in i/o bufs, then when pushing to i/o bufs, if they are nested we need to delay the request, otherwise we could send immediately.

pivotal-ning-yu replied:
or let i/o bufs support nest calls.

pivotal-ning-yu replied:
I plan to have a try on this.  The delayed delivery logic is ugly and complicated, getting rid of it could simplify the code significantly.

pivotal-ning-yu replied:
> I plan to have a try on this. The delayed delivery logic is ugly and complicated, getting rid of it could simplify the code significantly.

I can't find time to work on this last week.  As it is only an optional change, I decide to mark it as a post-pr task.  In current PR I'll focus on resolve review comments, not refactor.

pivotal-ning-yu commented:
Finished the README and comments.

zhangh43 commented on src/backend/cdb/motion/README.ic-proxy.md:
```diff
@@ -1,10 +1,62 @@
+
+It is recommended to setup this GUC with a script like below:
+
+    #!/usr/bin/env bash
```
This script make sense to developer.
How does customer configure this GUC? 
If still using this script, we need to handle bad case like the port(after -5000) already occupied by other program etc.
-5000 should also be an arg. And supply it as a management tool like gpstart?

pivotal-ning-yu replied:
In fact I should not say "recommended", I should only present this script as an example.

To make it a formal tool there are more things to do:
- make the port transformation configurable, as mentioned by you;
- make the ip resolution configurable, or even better, let the proxy code supports hostname directly;

pivotal-ning-yu commented:
I'll remove the pipeline changes from this PR, so we could have the PR in even before the libuv is packaged.  Once the libuv is packaged we could open a separate PR to enable the pipeline.

pivotal-ning-yu commented:
Force-pushed, removed the pipeline-related changes; squashed some of the commits; also rebased to the latest master to resolve conflicts.

pivotal-ning-yu commented:
Thanks for the review, I'll merge it now.  Next I'll work on a follow-up PR to enable the ic-proxy on the pipeline.
